HADOOP-16769 LocalDirAllocator to provide diagnostics when file creat…

ramesh0201 opened the pull request.

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  25m 22s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  19m 13s |  trunk passed  |
| +1 :green_heart: |  compile  |  16m 49s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 51s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 30s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 37s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  3s |  trunk passed  |
| +0 :ok: |  spotbugs  |   2m  6s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   2m  5s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 50s |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 10s |  the patch passed  |
| +1 :green_heart: |  javac  |  16m 10s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 52s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 26s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  14m 10s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  3s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   2m 14s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   9m 10s |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 53s |  The patch does not generate ASF License warnings.  |
|  |   | 131m 26s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.7 Server=19.03.7 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1892/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1892 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 9b5ef09f611d 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / cf9cf83 |
| Default Java | 1.8.0_242 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1892/1/testReport/ |
| Max. process+thread count | 3045 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1892/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/LocalDirAllocator.java:
```diff
@@ -441,9 +443,16 @@ public Path getLocalPathForWrite(String pathStr, long size,
+              returnPath = createPath(ctx.localDirs[dirNum], pathStr,
+                  checkWrite);
+            } catch (Exception e) {
+              errorText = e.getMessage();
```
1. Log the exception @ debug
2. store the caught exception  in a variable alongside errorText

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/LocalDirAllocator.java:
```diff
@@ -459,8 +468,13 @@ public Path getLocalPathForWrite(String pathStr, long size,
+      if (errorText != null) {
+        newErrorText = newErrorText + " due to " + errorText;
+      }
+      throw new DiskErrorException(newErrorText);
```
if any exception had been caught and stored in L484, use it in the constructor/setCause here. Stack traces are too important to lose.

steveloughran commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/TestLocalDirAllocator.java:
```diff
@@ -532,4 +533,19 @@ public void testGetLocalPathForWriteForInvalidPaths() throws Exception {
+    String dir0 = buildBufferDir(ROOT, 0);
+    String dir1 = buildBufferDir(ROOT, 1);
+    conf.set(CONTEXT, dir0 + "," + dir1);
+    LambdaTestUtils.intercept(DiskErrorException.class, "as the max capacity" +
```
minor nit, move the "as the max" capacity string onto a line on its own' merge it back to a single string,

ramesh0201 commented:
Thank you for the comments. For some reason I am not getting any email notifications about these git PR review comments. I just manually verified and found I missed to address these. I have addressed the comments and updated the PR. Can you please help me confirm the changes are expected?

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  33m 39s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| -1 :x: |  mvninstall  |   5m 15s |  root in trunk failed.  |
| +1 :green_heart: |  compile  |  23m 39s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 52s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 42s |  trunk passed  |
| -1 :x: |  shadedclient  |  21m 10s |  branch has errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  1s |  trunk passed  |
| +0 :ok: |  spotbugs  |   2m 13s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   2m 11s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 51s |  the patch passed  |
| +1 :green_heart: |  compile  |  17m 50s |  the patch passed  |
| +1 :green_heart: |  javac  |  17m 50s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 46s |  hadoop-common-project/hadoop-common: The patch generated 1 new + 39 unchanged - 0 fixed = 40 total (was 39)  |
| +1 :green_heart: |  mvnsite  |   1m 25s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| -1 :x: |  shadedclient  |  17m 45s |  patch has errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 56s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   2m 21s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   9m 42s |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 49s |  The patch does not generate ASF License warnings.  |
|  |   | 142m 32s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1892/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1892 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 6c34e90db15b 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / d08b9e94e36 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1892/2/artifact/out/branch-mvninstall-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1892/2/artifact/out/diff-checkstyle-hadoop-common-project_hadoop-common.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1892/2/testReport/ |
| Max. process+thread count | 1457 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1892/2/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 34s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  1s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  23m 32s |  trunk passed  |
| +1 :green_heart: |  compile  |  25m  3s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  compile  |  19m 42s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   0m 47s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 26s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m  8s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 36s |  hadoop-common in trunk failed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.  |
| +1 :green_heart: |  javadoc  |   0m 56s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |   2m 29s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   2m 27s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 54s |  the patch passed  |
| +1 :green_heart: |  compile  |  19m 51s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javac  |  19m 51s |  the patch passed  |
| +1 :green_heart: |  compile  |  18m 30s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  javac  |  18m 30s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 48s |  hadoop-common-project/hadoop-common: The patch generated 1 new + 39 unchanged - 0 fixed = 40 total (was 39)  |
| +1 :green_heart: |  mvnsite  |   1m 22s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  14m 22s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 43s |  hadoop-common in the patch failed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.  |
| +1 :green_heart: |  javadoc  |   1m  6s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  findbugs  |   2m 46s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  10m 31s |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 53s |  The patch does not generate ASF License warnings.  |
|  |   | 165m 10s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-1892/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1892 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 072e291fbc9e 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e756fe35909 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-1892/3/artifact/out/branch-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.txt |
| checkstyle | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-1892/3/artifact/out/diff-checkstyle-hadoop-common-project_hadoop-common.txt |
| javadoc | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-1892/3/artifact/out/patch-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.txt |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-1892/3/testReport/ |
| Max. process+thread count | 3046 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-1892/3/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

steveloughran commented:
Sorry, I'd completely missed this too.

@ramesh0201  -can you do a rebase and force push? I'd like to see what the checkstyle warnings were.

Other than the checkstyle issues, the patch LGTM & is almost ready to go in

hotcodemacha commented:
Thanks @ramesh0201 for the PR. Wanted to check if you are still working on it or I can take it forward.

ramesh0201 commented:
@ashutoshcipher Please feel free to take it forward
